store acceptable classnames in file

tatchi opened the pull request:
Still would like to figure out something but it should already be in a usable state.

dylanirlbeck commented:
Nice! I'll pull this down later this evening and take a look. Super excited to get this feature merged.

dylanirlbeck commented on lib/utils.re:
```diff
@@ -79,47 +73,37 @@ let getClassesFromSelector = selector => {
-        }
-      // Ignore other prelude
-      | _ => existingClassNames
+let getAcceptableClassNames = (cssRules, css) => {
```
Doesn't look like the `css` argument is used anymore

```suggestion
let getAcceptableClassNames = cssRules => {
```

tatchi replied:
Indeed, thx!

dylanirlbeck commented on lib/utils.re:
```diff
@@ -164,8 +145,57 @@ let checkAcceptable = (classNames, loc, tailwindFile) => {
+  acceptableClassNames: StringSet.t,
+};
+
+let getCachedAcceptaleClassNames = (~filename, ~tailwindFileContent) =>
```
**Typo:** `getCachedAcceptaleClassNames` -> `getCachedAcceptableClassNames`

tatchi replied:
done :)

dylanirlbeck commented on lib/utils.re:
```diff
@@ -164,8 +145,57 @@ let checkAcceptable = (classNames, loc, tailwindFile) => {
+
+let acceptableClassNames = ref(None);
+let validate = (~classNames, ~loc, ~tailwindFileContent) => {
+  let filename = "activeClassNames.bin";
```
**Question:** I want to confirm, is this the file name of `tailwind-ppx`'s cache? If possible, I'd prefer to follow the pattern set by [graphql-ppx](https://github.com/reasonml-community/graphql_ppx#ignore-graphql_ppx_cache-in-your-version-control) and name the file something like `.tailwind_ppx_cache`

tatchi commented:
Thanks for the review :)

> 2. Did the pre-commit `refmt` hook work for you? I configured it recently and was curious.

I didn't pay too much attention but haven't seen anything special in the console so I'm not sure it worked.

The active classnames are now cached in a `.tailwind_ppx_cache` directory (similarly to what `graphql_ppx` does). When running the ppx, the current directory is in `lib/bs` folder. I don't know if it's a good place to store this file. From what I understood, `graphql_ppx` [look higher in the hierarchy](https://github.com/reasonml-community/graphql_ppx/blob/master/src/base/read_schema.re#L285) up until they find the `graphql_schema.json` file (or whatever it's called) and store the cache directory at the same level.

We don't have such a file, so for now I stored the cache folder 2 steps higher `../../` which is where I think it's the best place to store it. The logic might be fragile though; an improvement might be to look up until we find a `bsconfig.json` file?

tatchi commented on lib/utils.re:
```diff
@@ -164,8 +146,94 @@ let checkAcceptable = (classNames, loc, tailwindFile) => {
+exception Ppx_cache_dir_is_not_dir;
+
+// Source: https://github.com/reasonml-community/graphql_ppx/blob/master/src/base/read_schema.re#L301
+let create_dir_if_not_exist = abs_path =>
```
I took this function from the `graphql_ppx` repo. For now errors are not handled but I guess we can improve that later on

dylanirlbeck replied:
Can we camelCase this (`createDirIfNotExist`) and the exception? I'd just like to be consistent with the rest of the codebase.

tatchi commented on test/native/lib/Test.re:
```diff
@@ -3,13 +3,6 @@ open Utils;
 open Css_types;
 open Helpers;
 
-let {describe, describeSkip, describeOnly} =
```
It produced errors so I just removed this. Not sure what it was used for but tests pass :D

dylanirlbeck replied:
This part was used to "clear the cache" of acceptableNames before each unit test was run, since the ref doesn't get reset until after all the tests have run. Since we're now doing actual caching logic, it kind of makes sense that the tests are now passing!

dylanirlbeck commented:
@tatchi I'm currently still reviewing this, but the pre-commit formatting hook is done with husky and lint-staged, so can you try running `yarn` in the project root and then committing a change to a Reason file again? It _should_ work after you've ran yarn to install dependencies.

tatchi commented on lib/utils.re:
```diff
@@ -164,8 +146,94 @@ let checkAcceptable = (classNames, loc, tailwindFile) => {
+      )
+    ) {
+    | 0 => Some(fileContent.acceptableClassNames)
+    | _ => None
```
Indeed, I missed this function when I looked at the doc. Thanks :)

tatchi commented:
I should have addressed the reviews.

> Just one idea about where we're putting the cache. I feel like it'd be best to always have the cache stored at the project root, since simply going two directories up might be a bit flaky. This will be especially important when we publish `tailwind-ppx` to opam, since the PPX may be in a different place.
> 
> Perhaps we could recursively look for an `esy.json` or `package.json`? Then we may have the problem if there's a misplaced `package.json` in some weird directory. In any event, **let's have this discussion in a separate PR**. I want to get this merged, and I think it works totally fine for normal BuckleScript projects.

I agree that looking for an `esy.json` or `package.json` file should be better (although probably not perfect). Let's indeed get this merged and improve that in another PR 😃
